server/sonar-server/src/main/java/org/sonar/ce/http/CeHttpClient.java

```diff
@@ -0,0 +1,146 @@
+  }
+
+  /**
+   * Connects to the specified JVM process and requests system information.
+   * @return the system info, or absent if the process is not up or if its HTTP URL
+   * is not registered into IPC.
+   */
+  public Optional<ProtobufSystemInfo.SystemInfo> retrieveSystemInfo() {
+    return call(SystemInfoActionClient.INSTANCE);
+  }
+
+  private enum SystemInfoActionClient implements ActionClient<Optional<ProtobufSystemInfo.SystemInfo>> {
+    INSTANCE;
+
+    @Override
+    public String getPath() {
+      return PATH_SYSTEM_INFO;
+    }
+
+    @Override
+    public Optional<ProtobufSystemInfo.SystemInfo> getDefault() {
+      return Optional.empty();
+    }
+
+    @Override
+    public Optional<ProtobufSystemInfo.SystemInfo> call(String url) throws Exception {
+      byte[] protobuf = IOUtils.toByteArray(new URI(url));
+      return Optional.of(ProtobufSystemInfo.SystemInfo.parseFrom(protobuf));
+    }
+  }
+
+  public void changeLogLevel(LoggerLevel level) {
+    call(new ChangeLogLevelActionClient(level));
+  }
+
+  private static final class ChangeLogLevelActionClient implements ActionClient<Void> {
+    private final LoggerLevel newLogLevel;
+
+    private ChangeLogLevelActionClient(LoggerLevel newLogLevel) {
+      this.newLogLevel = newLogLevel;
+    }
+
+    @Override
+    public String getPath() {
+      return PATH_CHANGE_LOG_LEVEL;
+    }
+
+    @Override
+    public Void getDefault() {
+      return null;
+    }
+
+    @Override
+    public Void call(String url) throws Exception {
+      okhttp3.Request request = new okhttp3.Request.Builder()
+        .post(RequestBody.create(null, new byte[0]))
+        .url(url + "?level=" + newLogLevel.name())
+        .build();
+      okhttp3.Response response = new OkHttpClient().newCall(request).execute();
+      if (response.code() != 200) {
+        Loggers.get(ChangeLogLevelAction.class).error(
+          "Failed to change log level in Compute Engine. Code was '{}' and response was '{}'",
+          response.code(),
+          response.body().string());
+      }
+      return null;
+    }
+  }
+
+  private <T> T call(ActionClient<T> actionClient) {
+    try (DefaultProcessCommands commands = DefaultProcessCommands.secondary(ipcSharedDir, COMPUTE_ENGINE.getIpcIndex())) {
+      if (commands.isUp()) {
+        return actionClient.call(commands.getHttpUrl() + "/" + actionClient.getPath());
+      }
+      return actionClient.getDefault();
+    } catch (Exception e) {
+      throw new IllegalStateException("Can not get system info of process " + COMPUTE_ENGINE, e);
+    }
+  }
+
+  private interface ActionClient<T> {
+    /**
+     * Path of the action.
+     */
+    String getPath();
+
+    /**
+     * Value to return when the Compute Engine is not ready.
+     */
+    T getDefault();
+
+    /**
+     * Delegates to perform the call to the Compute Engine's specified absolute URL.
+     */
+    T call(String url) throws Exception;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00112)